### ISSUE TITLE
Create scene inheritance for existing scenes

### ISSUE THREAD

**Issue by uapaladin** [godotengine/godot-proposals#1366]
I would like to have the option of making already existing scenes inherit from other scenes, either breaking up their current inheritance, or create one.

Please also see this issue where it was previously discussed:
[Issue 30024](https://github.com/godotengine/godot/issues/30024)

**Describe the project you are working on:**
I'm working on a 2D strategy game with many types of scenes with accompanying scripts, and most of them inherit from each other.

**Describe the problem or limitation you are having in your project:**
When needing to re-design/rework some class to inherit from a new intermediate class, there's not an option in the menu for this. There's an option 'New Inherited Scene' in the Scene menu, but this doesn't help. It's for creating new scenes. I need an option which does the same for already existing scenes. Now you will have to do it manually which ends up in having to edit scene files manually in an external text editor.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
An option for changing the inheritance of already existing scenes. When selected, you would point out the new inherited scene and the Godot system would make any necessary alterations in script and scene files. It helps by removing the necessity of manual editing of scene files.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
A new menu option in the right-click menu of scene files, in the FileSystem tab/window. It may have the caption 'Inherit from scene' or similar. When selecting it, you should get a screen/window from which you select the new base scene for the clicked-on scene file. This base scene should already exist. Any current inheritance of the clicked-on scene should switch to the new selected base scene.

Or alternatively, there could be two options:
1. Inherit from existing scene (in which you select the new base scene as described above)
2. Inherit from new scene (in which a new scene is created and set as the base scene for the one on which the operation is conducted)

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No. The only workaround involves manually editing scene files.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Yes. It would mean adding a new option to the menu system, which action concerns basic scene management.

**Comment by jonbonazza** [comment]
This has also been a pain for us in our game as well. Would love to see this supported.

**Comment by marcelb** [comment]
It's common in game development to avoid inheritance as much as possible, but it does make sense in some cases. I couldn't believe this isn't supported, when I first searched for this feature. In a long inheritance chain this creates a lot of rework, when you change the inheritance of a scene early in this chain.

There is a workaround though without ninja editing. 

Let's say we have this inheritance: A->B->C
Also let's say we want B to be the child of a new parent named D.

Without this feature you can't change B's inheritance from A, without completely recreating it, which without tinkering also leaves C hanging without inheritance, especially if it should have a different name (otherwise you could just overwrite the old file). This would need the dev to even recreate C.

You can do the following:
1) Save and close the editor of the scene (B) you want to change the inheritance for
2) Click "New Inherited Scene" and inherit D, name it the same way as B, but don't save it yet.
3) Right click the unsaved scenes root, select "Merge From Scene" and merge the whole scene B into it.
4) Drag'n'drop B's nodes into the new B's root and delete everything that was inherited (this is of course some work, depending on the size of B's node tree)
5) Save B and overwrite the old B.
6) Done.

This also works for adding inheritance when there was none before. But it keeps you from recreating the whole inheritance chain.

This of course is not an easy solution, it would make way more sense to have "Change Type" be able to inherit another scene instead of just built-in types, like in other game engines. I guess noone created this feature yet, because of name collisions that could happen that way, that needs to be solved? Like if you start inheriting a scene that already as a node with the same name as your child scene. You couldn't do the common enumerating approach, because that could break dependencies.

I also found it very odd, that you can't make an instanced scene root (which could've been a workaround as well) with the "Make Scene Root" command, with an error that says "Instantiated scenes can't become root." which is clearly false, because you can create it with "New inherited scene". It's not logical atm.

**Comment by DissonantVoid** [comment]
looks like the workaround is no longer viable in Godot 4 since "Merge From Scene" is no longer available. is there a way to work around this issue in Godot 4 without having to recreat every scene that derives from the scene I'm trying to change the base scene of?

**Comment by bsil78** [comment]
I was myself disappointed to not find Merge from scene anymore... this was an effective refactoring help. A similar refactoring help feature would be very appreciated as it brings frustration in commonly praticed workflows.

**Comment by bsil78** [comment]
He was right then : https://github.com/godotengine/godot/pull/45943#pullrequestreview-593256815

**Comment by klootas** [comment]
Yes, having a "Clear Inheritance" context menu item but no "Update Inheritance" just feels like teasing. It would be awesome to be able to update a scene's base class without hacks 👍
